### PR TITLE
volumes: fix error-handling when removing volumes with swarm enabled

### LINKED ISSUE #4082: Attempting to Remove Volume Used by Service Does Not Report Error with CLI 23.0.1
### Description

When you attempt to remove a volume that's currently in use by a service, you should get an error stating that the volume is in use. With 23.0.1 you do not get an error.

The service usage is my primary scenario, but the same thing happens when using just a container without a service. 

Inspecting the container does show it as using the volume.

I came across this in some tests I have for code that wraps the docker CLI. I'm testing that my code properly passes back the error from docker. So just a coincidence I'm trying this scenario really.

### Reproduce

1. `docker volume create TestVolume`
```
TestVolume
```
2. `docker service create --name TestService --mount type=volume,source=TestVolume,destination=/volume busybox:1.34.1 sleep infinity`
```
yef18ac5rllrhib1wmpk86r22
overall progress: 1 out of 1 tasks
1/1: running   [==================================================>]
verify: Service converged
```
3. `docker volume rm TestVolume` (this should produce an error)
```
TestVolume
```
4. `docker volume ls`
```
DRIVER    VOLUME NAME
local     TestVolume
```

### Expected behavior

When the same operation is done with the Docker CLI 20.10.22 you get this:

` docker volume rm TestVolume`
```
Error response from daemon: remove TestVolume: volume is in use - [85c3c0bea04f14ef8696a038ff1f5f08fd796dbff61276459c9c4336867a2072]
```

### docker version

```bash
Client:
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:50:24 2023
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:46:47 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.1
    Path:     C:\Users\twinter\.docker\cli-plugins\docker-sbom.exe

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 20
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: 1bis6jp8bw0eg3x73wfrq1px5
  Is Manager: true
  ClusterID: exugqkpkx2cn5bp38wcw73dxo
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.2.96.83
  Manager Addresses:
   10.2.96.83:2377
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 4.18.0-425.3.1.el8.x86_64
 Operating System: Red Hat Enterprise Linux 8.7 (Ootpa)
 OSType: linux
 Architecture: x86_64
 CPUs: 3
 Total Memory: 5.741GiB
 Name: PRD-VIRT-RHL-56.AD.VERVEINDUSTRIAL.COM
 ID: IPGK:JLNH:EKXO:NZLH:O2O4:GQC5:PE7C:NYBC:BKQA:HNR2:Q36F:WCJ5
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### LINKED ISSUE #45037: Docker volume rm
### Description

Removing a volume which is in use by a container silently fails instead of throwing an error (i.e. HTTP 409). This was working as expected in 20.10.x versions but stopped working in 23.0.1.

### Reproduce

1. docker volume create foo
2. docker container create -v foo:/data busybox:latest
3. docker volume rm foo # check exit status, it shouldn't be 0

### Expected behavior

`docker volume rm` should report failure if volume is in use

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:46:54 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:46:54 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 2
  Running: 0
  Paused: 0
  Stopped: 2
 Images: 36
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: c2bypvx3dlonfugudvoihdcaa
  Is Manager: true
  ClusterID: tbeiz6qhtjpvuz1bvx6104q3g
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: true
  Root Rotation In Progress: false
  Node Address: 127.0.0.1
  Manager Addresses:
   127.0.0.1:2377
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.0-3-amd64
 Operating System: Debian GNU/Linux bookworm/sid
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 14.98GiB
 Name: sourtoukis
 ID: N6UF:HU6Y:6LLD:NA4I:OXAO:WQPK:5THP:K3XT:KUAU:54AQ:QNTB:7XCE
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 38
  Goroutines: 161
  System Time: 2023-02-19T16:44:49.979529486+02:00
  EventsListeners: 0
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by thaJeztah**
- relates to https://github.com/moby/moby/pull/44224
- fixes https://github.com/docker/cli/issues/4082
- fixes https://github.com/moby/moby/issues/45037

### integration/volumes: TestVolumesRemove: add coverage for force/no-force

Add additional test-cases for deleting non-existing volumes (with/without force).

With this patch:

    make TEST_FILTER=TestVolumesRemove DOCKER_GRAPHDRIVER=vfs test-integration

    Running /go/src/github.com/docker/docker/integration/volume (arm64.integration.volume) flags=-test.v -test.timeout=10m  -test.run TestVolumesRemove
    ...
    === RUN   TestVolumesRemove
    === RUN   TestVolumesRemove/volume_in_use
    === RUN   TestVolumesRemove/volume_not_in_use
    === RUN   TestVolumesRemove/non-existing_volume
    === RUN   TestVolumesRemove/non-existing_volume_force
    --- PASS: TestVolumesRemove (0.04s)
        --- PASS: TestVolumesRemove/volume_in_use (0.00s)
        --- PASS: TestVolumesRemove/volume_not_in_use (0.01s)
        --- PASS: TestVolumesRemove/non-existing_volume (0.00s)
        --- PASS: TestVolumesRemove/non-existing_volume_force (0.00s)
    PASS

### volumes: fix error-handling when removing volumes with swarm enabled

Commit 3246db3755698455740c812c6477f43e74924fa7 (https://github.com/moby/moby/pull/44224) added handling for removing
cluster volumes, but in some conditions, this resulted in errors not being
returned if the volume was in use;

    docker swarm init
    docker volume create foo
    docker create -v foo:/foo busybox top
    docker volume rm foo

This patch changes the logic for ignoring "local" volume errors if swarm
is enabled (and cluster volumes supported).

While working on this fix, I also discovered that Cluster.RemoveVolume()
did not handle the "force" option correctly; while swarm correctly handled
these, the cluster backend performs a lookup of the volume first (to obtain
its ID), which would fail if the volume didn't exist.

Before this patch:

    make TEST_FILTER=TestVolumesRemoveSwarmEnabled DOCKER_GRAPHDRIVER=vfs test-integration
    ...
    Running /go/src/github.com/docker/docker/integration/volume (arm64.integration.volume) flags=-test.v -test.timeout=10m  -test.run TestVolumesRemoveSwarmEnabled
    ...
    === RUN   TestVolumesRemoveSwarmEnabled
    === PAUSE TestVolumesRemoveSwarmEnabled
    === CONT  TestVolumesRemoveSwarmEnabled
    === RUN   TestVolumesRemoveSwarmEnabled/volume_in_use
        volume_test.go:122: assertion failed: error is nil, not errdefs.IsConflict
        volume_test.go:123: assertion failed: expected an error, got nil
    === RUN   TestVolumesRemoveSwarmEnabled/volume_not_in_use
    === RUN   TestVolumesRemoveSwarmEnabled/non-existing_volume
    === RUN   TestVolumesRemoveSwarmEnabled/non-existing_volume_force
        volume_test.go:143: assertion failed: error is not nil: Error response from daemon: volume no_such_volume not found
    --- FAIL: TestVolumesRemoveSwarmEnabled (1.57s)
        --- FAIL: TestVolumesRemoveSwarmEnabled/volume_in_use (0.00s)
        --- PASS: TestVolumesRemoveSwarmEnabled/volume_not_in_use (0.01s)
        --- PASS: TestVolumesRemoveSwarmEnabled/non-existing_volume (0.00s)
        --- FAIL: TestVolumesRemoveSwarmEnabled/non-existing_volume_force (0.00s)
    FAIL

With this patch:

    make TEST_FILTER=TestVolumesRemoveSwarmEnabled DOCKER_GRAPHDRIVER=vfs test-integration
    ...
    Running /go/src/github.com/docker/docker/integration/volume (arm64.integration.volume) flags=-test.v -test.timeout=10m  -test.run TestVolumesRemoveSwarmEnabled
    ...
    === RUN   TestVolumesRemoveSwarmEnabled
    === PAUSE TestVolumesRemoveSwarmEnabled
    === CONT  TestVolumesRemoveSwarmEnabled
    === RUN   TestVolumesRemoveSwarmEnabled/volume_in_use
    === RUN   TestVolumesRemoveSwarmEnabled/volume_not_in_use
    === RUN   TestVolumesRemoveSwarmEnabled/non-existing_volume
    === RUN   TestVolumesRemoveSwarmEnabled/non-existing_volume_force
    --- PASS: TestVolumesRemoveSwarmEnabled (1.53s)
        --- PASS: TestVolumesRemoveSwarmEnabled/volume_in_use (0.00s)
        --- PASS: TestVolumesRemoveSwarmEnabled/volume_not_in_use (0.01s)
        --- PASS: TestVolumesRemoveSwarmEnabled/non-existing_volume (0.00s)
        --- PASS: TestVolumesRemoveSwarmEnabled/non-existing_volume_force (0.00s)
    PASS

**- Description for the changelog**

```
Fix removing volumes not returning an error if swarm is enabled
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Hm... looks like some other tests are now failing;

```
=== Failed
=== FAIL: api/server/router/volume TestVolumeRemoveSwarm (0.00s)
    volume_routes_test.go:456: assertion failed: error is not nil: volume vol1 not found

=== FAIL: api/server/router/volume TestVolumeRemoveSwarmForce (0.00s)
    volume_routes_test.go:567: assertion failed: expression is false: errdefs.IsConflict(err)
```

File: api/server/router/volume/volume_routes.go [hunk 1]
```diff
@@ -159,25 +159,35 @@ func (v *volumeRouter) deleteVolumes(ctx context.Context, w http.ResponseWriter,
...
-	if err != nil || force {
+	if err != nil && !errdefs.IsNotFound(err) {
+		return err
+	}
+
+	// If no volume was found, the volume may be a cluster volume, or force
+	// is enabled, in which case the remove won't return an error if no local
+	// volume existed, so we don't know if removal succeeded. In that case
+	// we always try to delete cluster volumes as well.
+	if errdefs.IsNotFound(err) || force {
```
Comment (thaJeztah) [inline api/server/router/volume/volume_routes.go:175]: Note that checking for `errdefs.IsNotFound()` is actually redundant here; I thought it made the code slightly more descriptive, but happy to revert that change.

File: api/server/router/volume/volume_routes.go [hunk 2]
```diff
@@ -159,25 +159,35 @@ func (v *volumeRouter) deleteVolumes(ctx context.Context, w http.ResponseWriter,
 	}
 	force := httputils.BoolValue(r, "force")
 
-	version := httputils.VersionFromContext(ctx)
-
+	// First we try deleting local volume. The volume may not be found as a
+	// local volume, but could be a cluster volume, so we ignore "not found"
+	// errors at this stage. Note that no "not found" error is produced if
+	// "force" is enabled.
 	err := v.backend.Remove(ctx, vars["name"], opts.WithPurgeOnError(force))
```
Comment (thaJeztah) [inline api/server/router/volume/volume_routes.go:166]: I'm also looking if we can remove the `force` from the backends, and instead handle "not-exist" errors on the caller site, but looks like that may need more work ("purge" does more than only ignoring "not-exist" errors, as it also updates state in the backend to remove volumes from the store if the (_remote_ ?) volume store (no longer) has a volume.

**Comment by thaJeztah** [comment]
@cpuguy83 @dperny @neersighted ptal

**Comment by dperny** [comment]
LGTM!

File: api/server/router/volume/volume_routes.go [hunk 3]
```diff
@@ -159,25 +159,35 @@ func (v *volumeRouter) deleteVolumes(ctx context.Context, w http.ResponseWriter,
...
-	// when a removal is forced, if the volume does not exist, no error will be
-	// returned. this means that to ensure forcing works on swarm volumes as
-	// well, we should always also force remove against the cluster.
-	if err != nil || force {
+	if err != nil && !errdefs.IsNotFound(err) {
+		return err
+	}
```
Comment (neersighted) [inline api/server/router/volume/volume_routes.go:169]: The conditional flow here (and through all of these blocks) feels very clunky. Is there a better way to structure this that makes the flow more obvious, even if we have mode code movement?

File: api/server/router/volume/volume_routes.go [hunk 4]
```diff
@@ -159,25 +159,35 @@ func (v *volumeRouter) deleteVolumes(ctx context.Context, w http.ResponseWriter,
...
+	// In that case we always try to delete cluster volumes as well.
+	if errdefs.IsNotFound(err) || force {
+		version := httputils.VersionFromContext(ctx)
 		if versions.GreaterThanOrEqualTo(version, clusterVolumesVersion) && v.cluster.IsManager() {
-			if errdefs.IsNotFound(err) || force {
-				err := v.cluster.RemoveVolume(vars["name"], force)
-				if err != nil {
-					return err
-				}
+			if cvErr := v.cluster.RemoveVolume(vars["name"], force); cvErr != nil {
```
Comment (cpuguy83) [inline api/server/router/volume/volume_routes.go:178]: Can we unify the branches here?
I think we can do `err = v.cluster.RemoveVolume` and then do nothing else in the branch.
This of course assumes `cluster.RemoveVolume` will return a not found error if `force=false`

File: integration/volume/volume_test.go [hunk 5]
```diff
@@ -75,16 +78,83 @@ func TestVolumesRemove(t *testing.T) {
...
+		})
+		assert.NilError(t, err)
+
+		err = client.VolumeRemove(ctx, vname, false)
+		assert.NilError(t, err)
+	})
+
+	t.Run("non-existing volume", func(t *testing.T) {
+		err = client.VolumeRemove(ctx, "no_such_volume", false)
+		assert.Check(t, is.ErrorType(err, errdefs.IsNotFound))
```
Comment (cpuguy83) [inline integration/volume/volume_test.go:151]: I'm not sure how this error check works?
`errdefs.IsNotFound` is a function

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: integration/volume/volume_test.go (hunk 5 above)
Comment (thaJeztah) [reply integration/volume/volume_test.go:151]: Ah, yes, so `is.ErrorType` supports either a type or an assertion function, which `errdefs.IsXXX` satisfies

(also didn't know this until not too long ago)

File: api/server/router/volume/volume_routes.go (hunk 3 above)
Comment (thaJeztah) [reply api/server/router/volume/volume_routes.go:169]: Yes, I was hoping to remove the "force" from the backend entirely, and just handle `notfound` errors (which are the ones we want to ignore) to the caller site; https://github.com/moby/moby/pull/45148#discussion_r1133268416

Unfortunately, there is code in the backend that still needs it (if the _plugin_ doesn't have the volume, then our local storage still has to remove it, and not fail)

_Perhaps_ still possible, if we would have the backend ignore the error, for what it needs to do, but make it still return it, but that still requires more changes (probably out of scope for this PR)

File: api/server/router/volume/volume_routes.go (hunk 4 above)
Comment (thaJeztah) [reply api/server/router/volume/volume_routes.go:178]: Oh! You're right; I guess that would work; just overwrite the original `err` value if we reach this code.

**Review by cpuguy83 — APPROVED** [review]

**Review by vvoland — APPROVED** [review]

**Comment by thaJeztah** [comment]
Let me bring this one in, and do a cherry-pick later

**Comment by thaJeztah** [comment]
cherry-pick in https://github.com/moby/moby/pull/45155

**Pull request merged**